src/backend/distributed/commands/create_citus_local_table.c

```diff
@@ -0,0 +1,911 @@
+							   "not exist")));
+	}
+
+	Oid relationId = relation->rd_id;
+
+	ErrorIfCoordinatorNotAddedAsWorkerNode();
+	ErrorIfUnsupportedCitusLocalTableKind(relationId);
+	EnsureTableNotDistributed(relationId);
+
+	/*
+	 * When creating other citus table types, we don't need to check that case as
+	 * EnsureTableNotDistributed already errors out if the given relation implies
+	 * a citus table. However, as we don't mark the relation as citus table, i.e we
+	 * do not use the relation with relationId as the shell relation, parallel
+	 * create_citus_local_table executions would not error out for that relation.
+	 * Hence we need to error out for shard relations too.
+	 */
+	ErrorIfRelationIsAKnownShard(relationId);
+
+	/*
+	 * We do not allow creating citus local table if the table is involved in a
+	 * foreign key relationship with "any other table". Note that we allow self
+	 * references.
+	 */
+	ErrorIfTableHasExternalForeignKeys(relationId);
+
+	/* we do not support policies in citus community */
+	ErrorIfUnsupportedPolicy(relation);
+}
+
+
+/*
+ * ErrorIfUnsupportedCitusLocalTableKind errors out if the relation kind of
+ * relation with relationId is not supported for citus local table creation.
+ */
+static void
+ErrorIfUnsupportedCitusLocalTableKind(Oid relationId)
+{
+	const char *relationName = get_rel_name(relationId);
+
+	if (IsChildTable(relationId) || IsParentTable(relationId))
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot create citus local table \"%s\", citus local "
+							   "tables cannot be involved in inheritance relationships",
+							   relationName)));
+	}
+
+	if (PartitionTable(relationId))
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot create citus local table \"%s\", citus local "
+							   "tables cannot be partition of other tables",
+							   relationName)));
+	}
+
+	char relationKind = get_rel_relkind(relationId);
+	if (!(relationKind == RELKIND_RELATION || relationKind == RELKIND_FOREIGN_TABLE))
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot create citus local table \"%s\", only regular "
+							   "tables and foreign tables are supported for citus local "
+							   "table creation", relationName)));
+	}
+}
+
+
+/*
+ * CreateCitusLocalTable is the internal method that creates a citus table
+ * from the table with relationId. The created table would have the following
+ * properties:
+ *  - it will have only one shard,
+ *  - its distribution method will be DISTRIBUTE_BY_NONE,
+ *  - its replication model will be ReplicationModel,
+ *  - its replication factor will be set to 1.
+ * Similar to reference tables, it has only 1 placement. In addition to that, that
+ * single placement is only allowed to be on the coordinator.
+ */
+static void
+CreateCitusLocalTable(Oid relationId)
+{
+	/* these checks should be done before acquiring the lock on the table */
```

Comment:
Minor: it could be nicer to explain "why" these checks should be done "before". That way when adding something new, it could help people place it to the right place